typed-lambda-calculus/article.md

```diff
@@ -0,0 +1,717 @@
+
+The most common argument for the strongly-typed languages is that if a program type checks, it will never segfault (or have any other type-related issues), but I think this argument sells typed languages short.
+A sufficiently strict and expressive type system allows for encoding much more than simply "not segfaulting".
+In some cases, the whole domain can be expressed in types.
+
+In this article, we're concerned with the basics, for the most part, so we won't get too much into the more expressive type systems like System FC or systems with dependent types.
+
+## Lambda calculus with atomic boolean values
+
+We will now introduce atomic values to the lambda calculus.
+While it is feasible to construct a value representation using only λ-abstractions (using Church encoding or otherwise) for almost any value we might be interested in, practically speaking, it's not very convenient.
+
+Thus, we introduce atomic values $true$ and $false$.
+
+Since we introduced atomic values, we also need to introduce an eliminator (i.e. a consumer) for these values.
+In this case let's consider an $\mathbf{if} \ldots \mathbf{then} \ldots \mathbf{else} \ldots$ operator.
+
+The semantics of this operator should be reasonably familiar to a practicing programmer, but let's informally write them them out anyway:
+
+- $\mathbf{if}\; true\; \mathbf{then}\;\alpha\;\mathbf{else} \;\beta$ should reduce to $\alpha,$
+- $\mathbf{if}\; false\; \mathbf{then}\;\alpha\;\mathbf{else} \;\beta$ should reduce to $\beta,$
+- $\mathbf{if}\; \gamma\; \mathbf{then}\;\alpha\;\mathbf{else} \;\beta$ where $\gamma$ is a redex, $\gamma$ should be reduced first.
+
+The introduction of the $\mathbf{if}$ operator creates a bit of a kink.
+What do we do if $\gamma$ is neither $true$, $false$, nor a redex?
+E.g. $\mathbf{if}\; (\lambda x. x)\; \mathbf{then}\;\ldots\;\mathbf{else} \;\ldots$
+
+Such terms are called "stuck", the computation, according to our rules, can not proceed.
+We want to avoid stuck terms!
+
+To do that, we'll have to place some restrictions on what $\gamma$ could be.
+
+### Typing relations, rules, and systems
+
+To express such restrictions, we introduce the so-called _typing relation_.
+A typing relation is a relation, in this case meaning a set of pairs, which assigns a _type_ to each and every expression in the language.
+
+One such individual assignment we'll call _typing judgement_, and will write it separated by a column: $t : T$, read as "term $t$ has type $T$".
+
+Types can be thought of as sets of terms of a certain general form.
+Turns out (Reynolds, 1984), this naive interpretation of types as sets doesn't hold to rigorous scrutiny.
+Still, it's a good enough mental model unless you're a type theory researcher (and if you are, hi, very nice of you to read this article, please get in touch if you have corrections).
+
+Typing relation thus assigns a type to each term.
```

Comment:
Same comment as above. It's not each term, it's only some of them (and those are well-typed). All other terms are ill-typed.